Add SchemaModel column name access through class attributes

jespercodes opened the pull request:
This adresses #364

jeffzi commented on tests/test_model.py:
```diff
@@ -386,6 +386,28 @@ class Child(Mid):
 
 
+def test_inherit_schemamodel_fields_alias():
+    """Test that columns and indices are inherited."""
```
Test that column and index **aliases** are inherited :)

jeffzi commented on tests/test_model.py:
```diff
@@ -566,3 +588,94 @@ class Child(Base):
+        b: Series[str]
+
+    schema = Child.to_schema()
+    print(schema.columns)
```
Remove print statement

jespercodes replied:
oops, my bad ;)

jeffzi commented on tests/test_model.py:
```diff
@@ -566,3 +588,94 @@ class Child(Base):
     assert schema.columns.get("_b", None) is not None
+
+
+def test_inherit_override():
```
We can merge this test with the one above `test_inherit_alias`.

jespercodes replied:
I tried to systematically cover all cases (omitted, present, alias) and their combinations in one test.

jeffzi reviewed (changes requested):
Thanks @jespercodes !  I've requested minor changes but overall awesome PR :rocket: 

`__init_subclass__` makes sure all fields have a `FieldInfo`. Therefore, I think:
*  `_getattr_fieldinfo` doesn't need to support a default. 
* We dont need to test for `if field is None` at `models.py#291`.

My reasoning is that all fields should be set, and if they are not it's a bug that the user can't do anything about. 

Code coverage should be up again after the requested changes, otherwise some tweaks to the tests can be added.

jespercodes commented on pandera/model_components.py:
```diff
@@ -72,6 +73,22 @@ def __init__(
+        else:
+            return self.original_name
+
+    def __set__(self, instance, value):
```
Absolutely, wanted it to be a data descriptor. Should be clear now

jespercodes commented on pandera/model_components.py:
```diff
@@ -103,6 +120,10 @@ def to_column(
             checks=checks,
         )
 
+    def column_name(self):
```
done

jespercodes commented on pandera/model.py:
```diff
@@ -75,6 +76,16 @@ def _extract_config_options(config: Type) -> Dict[str, Any]:
+    for class_ in cls.__mro__:
+        if name in class_.__dict__:
+            return class_.__dict__[name]
+    return default
```
Changed the mro call. I did leave a `return None` at the end. Was considering raising an error, but it seems like all sensible cases are already covered with existing tests with good error messages for how the user should change their class definition to be correct.

jeffzi replied:
I commited changes (see my reply in main thread). Let me know if you disagree.

jespercodes commented:
Thanks @jeffzi for clear and actionable feedback. I've tried to adress all points. The one thing I'm not sure how to handle cleanly is the "this part should never be reached, if so it is a bug" in `_getattr_fieldinfo` function and where it is used. Maybe the intent can be expressed more clearly, without having to use pragmas for codecov. Let me know what you think about the current situation.

Maybe some tweaks to the documentation pages would be appropriate as a next step?

Also, should I convert this PR to 'draft', does it matter for you?

jeffzi commented:
>  The one thing I'm not sure how to handle cleanly is the "this part should never be reached, if so it is a bug" in _getattr_fieldinfo function and where it is used.

I took the liberty of refactoring the part that pulls the attributes. Specifically:
1.  I refactored `_getattr_fieldinfo` into `get_model_attrs`. It gathers **all** the attributes similarly to inspect.get_members, except that it bypasses the descriptors `__get__`. Because of `init_subclass`, we know all annotations must be present in the attributes.
2.  I merged  `_collect_field_annotations()` and `collect_fields()` so that we can walk the mro only once to collect the fields.

> Also, should I convert this PR to 'draft', does it matter for you?

I usually reserve "draft" for prototypes. Your PR will be merged for sure, no need to make it a draft :)

> Maybe some tweaks to the documentation pages would be appropriate as a next step?

Yeah, I forgot to mention that ! The doc is located [here](https://github.com/pandera-dev/pandera/blob/master/docs/source/schema_models.rst). Here are good places to start:
* Add a subsection for the feature: explanation + add an example of selecting a DataFrame with the schema's columns.
* [Checks](https://github.com/pandera-dev/pandera/blob/master/docs/source/schema_models.rst#columnindex-checks): tweak one case to use descriptor. 
* [Aliases](https://github.com/pandera-dev/pandera/blob/master/docs/source/schema_models.rst#aliases): tweak example to show that descriptors respect aliases.

Let me know if you need help with the documentation or anything else.

jespercodes commented:
> I took the liberty of refactoring the part that pulls the attributes. Specifically:

Very nice! I see from the python docs that [`get_type_hints`](https://docs.python.org/3/library/typing.html#typing.get_type_hints) actually walks the mro. All good stuff to avoid calling that mutliple times 👍 Also looks alot cleaner now, great! :)

> Add a subsection for the feature: explanation + add an example of selecting a DataFrame with the schema's columns.

Maybe just have one single example under the "Basic Usage" section is enough?

> Aliases: tweak example to show that descriptors respect aliases.

Seems straightforward to add, will do!

Will try to do these above in the next day or so.

> Checks: tweak one case to use descriptor.

However for Checks, we might want to discuss how to handle it. I used the example in the docs and made three different tweaks to show how it works with the most recent commit.

1) Referencing the class within the class is not viable. The class is not yet created when we want to use it:

```python
import pandas as pd
import pandera as pa
from pandera.typing import Index, Series

class CustomCheckSchema(pa.SchemaModel):

    a: Series[int] = pa.Field(gt=0, coerce=True)
    abc: Series[int]
    idx: Index[str]

    @pa.check(CustomCheckSchema.a, name="foobar")
    def custom_check(cls, a: Series[int]) -> Series[bool]:
        return a < 100

    @pa.check("^a", regex=True, name="foobar")
    def custom_check_regex(cls, a: Series[int]) -> Series[bool]:
        return a > 0

    @pa.check("idx")
    def check_idx(cls, idx: Index[int]) -> Series[bool]:
        return idx.str.contains("dog")
#> Traceback (most recent call last):
#> <ipython-input-1-e5d2dc985829> in <module>
#> ----> 1 class CustomCheckSchema(pa.SchemaModel):
#>       2 
#>       3     a: Series[int] = pa.Field(gt=0, coerce=True)
#>       4     abc: Series[int]
#>       5     idx: Index[str]
#> <ipython-input-1-e5d2dc985829> in CustomCheckSchema()
#>       5     idx: Index[str]
#>       6 
#> ----> 7     @pa.check(CustomCheckSchema.a, name="foobar")
#>       8     def custom_check(cls, a: Series[int]) -> Series[bool]:
#>       9         return a < 100
#> NameError: name 'CustomCheckSchema' is not defined
```

<sup>Created on 2021-01-18 by the [reprexpy package](https://github.com/crew102/reprexpy)</sup>

2) Using the variable is the right way, but it does not currently work. I think we need to handle the `FieldInfo` case within `pa.check`. I have not investigated this thoroughly yet, but I guess the descriptor is not invoked when referenced like this within the class. We just pass it the instance directly

```python
# %%
import pandas as pd
import pandera as pa
from pandera.typing import Index, Series

class CustomCheckSchema(pa.SchemaModel):

    a: Series[int] = pa.Field(gt=0, coerce=True)
    abc: Series[int]
    idx: Index[str]

    @pa.check(a, name="foobar")
    def custom_check(cls, a: Series[int]) -> Series[bool]:
        return a < 100

    @pa.check("^a", regex=True, name="foobar")
    def custom_check_regex(cls, a: Series[int]) -> Series[bool]:
        return a > 0

    @pa.check("idx")
    def check_idx(cls, idx: Index[int]) -> Series[bool]:
        return idx.str.contains("dog")

schema = CustomCheckSchema.to_schema()
#> Traceback (most recent call last):
#> <ipython-input-1-90288546d15e> in <module>
#> ----> 1 schema = CustomCheckSchema.to_schema()
#> ~/Repos/pandera-featstr/pandera/pandera/model.py in to_schema(cls)
#>     121             field.name for (_, (_, field)) in cls.__fields__.items()
#>     122         ]
#> --> 123         cls.__checks__ = cls._extract_checks(
#>     124             check_infos, field_names=column_names
#>     125         )
#> ~/Repos/pandera-featstr/pandera/pandera/model.py in _extract_checks(cls, check_infos, field_names)
#>     337             for field in matched:
#>     338                 if field not in field_names:
#> --> 339                     raise SchemaInitError(
#>     340                         f"Check {check_.name} is assigned to a non-existing field '{field}'."
#>     341                     )
#> SchemaInitError: Check foobar is assigned to a non-existing field '<pandera.model_components.FieldInfo object at 0x7fbf7c4110a0>'.
```

<sup>Created on 2021-01-18 by the [reprexpy package](https://github.com/crew102/reprexpy)</sup>

3) When a `pa.Field()` is omitted, we don't actually define/set the variable, so it cannot be used within the class since `__init_subclass__` has not been called yet.

```python
import pandas as pd
import pandera as pa
from pandera.typing import Index, Series

class CustomCheckSchema(pa.SchemaModel):

    a: Series[int] = pa.Field(gt=0, coerce=True)
    abc: Series[int]
    idx: Index[str]

    @pa.check("a", name="foobar")
    def custom_check(cls, a: Series[int]) -> Series[bool]:
        return a < 100

    @pa.check("^a", regex=True, name="foobar")
    def custom_check_regex(cls, a: Series[int]) -> Series[bool]:
        return a > 0

    @pa.check(idx)
    def check_idx(cls, idx: Index[int]) -> Series[bool]:
        return idx.str.contains("dog")
#> Traceback (most recent call last):
#> <ipython-input-1-7a3ee6a71944> in <module>
#> ----> 1 class CustomCheckSchema(pa.SchemaModel):
#>       2 
#>       3     a: Series[int] = pa.Field(gt=0, coerce=True)
#>       4     abc: Series[int]
#>       5     idx: Index[str]
#> <ipython-input-1-7a3ee6a71944> in CustomCheckSchema()
#>      13         return a > 0
#>      14 
#> ---> 15     @pa.check(idx)
#>      16     def check_idx(cls, idx: Index[int]) -> Series[bool]:
#>      17         return idx.str.contains("dog")
#> NameError: name 'idx' is not defined
```

<sup>Created on 2021-01-18 by the [reprexpy package](https://github.com/crew102/reprexpy)</sup>

jespercodes commented:
Thought a bit more about case 2) above. For what I know the "parts of the class" are resolved in the following order on class definition (basically read top to bottom, left to right):

Given the base case in the docs without any tweaks above
1) a = pa.Field() with its `__init__` is called
2) the methods on the classes are created, and with that actually calling `pa.check` for each one
3) The class is now done, python calls __set_name__ on `a`
4) `__init_subclass__` is called, creating `FieldInfo`s and manually  invoking `__set_name__` on the created  for omitted annotations


I guess we have to resolve any potential `FieldCheckInfo.fields` which are not strings, in `to_schema()`, when we know everything is setup correctly.

Can look a bit further into this and see if I can get it to work, but it might also be scoped to another issue/PR. Let me know

jeffzi commented:
> I see from the python docs that get_type_hints actually walks the mro. 

True. Actually, we walk the mro again to build the `Config` and the `checks`. Ideally, we would want to do everything at the same time. Pydantic does [that](https://github.com/samuelcolvin/pydantic/blob/13a5c7d676167b415080de5e6e6a74bea095b239/pydantic/main.py#L212) in a metaclass`.__new__`, along with a custom `get_type_hints`. However, optimizing model initialization makes sense for pydantic since it creates many more models than pandera does. I assume people dont have crazy amounts of schemas. That said, the current implementation of `SchemaModel` is quite clean and I'd rather focus on features for now :)

> Maybe just have one single example under the "Basic Usage" section is enough?

Totally :+1: Maybe add just one short comment since it's not a standard behavior.


Regarding checks:
1. Makes sense, my bad for suggesting this use-case.
3. We can't do anything about it since python can't even resolve the name. We could however warn the user in the doc.

> I guess we have to resolve any potential FieldCheckInfo.fields which are not strings, in to_schema(), when we know everything is setup correctly.
2. I agree with your analysis. `model_components.FieldCheckInfo.fields` holds the first argument of `pa.check`. Now it can contain `FieldInfo`s. Since case 3 is impossible, we know that `pa.Field` was not omitted and therefore `FieldInfo.name`  is ready.  Imho, the best place to resolve the name is `SchemaModel._extract_checks`.  

> Can look a bit further into this and see if I can get it to work, but it might also be scoped to another issue/PR. 

Based on our discussion, it should be a minor change. Let's keep it here and open another issue if the scope proves to be largerr than expected.

Thanks for your perseverance :rocket:

jespercodes commented:
I seemed to have snucked in a regression in one of the early commits reagarding `pa.check` on field with alias. I added some tweaks to the test suite to cover that and fixed it.

Also added the docs, let me know what you think @jeffzi . In my mind this should be real close now to be able to ship 🚀

jeffzi commented:
I added a note about accessing descriptor from class scope. 

I think we're good ! Thanks @jespercodes for this first PR.

cosmicBboy commented:
thanks @jespercodes @jeffzi, this is awesome!

I'm gonna start a `dev` branch so that we can merge changes for major/minor versions into that instead of `master`. This'll allow us to do `bugfix` releases in between major/minor releases.

GiorgioBalestrieri commented:
FWIW, I'm exploring how I could use `pandera` to improve data validation, and this feature is awesome. I had implemented (with others) a very scrappy validation tool for dataframes, and being able to rely on field names instead of strings (it's much less error prone) is important to our use cases. Thanks!
